plugins/parquet-store/parquet_store.cpp

```diff
@@ -0,0 +1,722 @@
+  self->state.self_ = self;
+  self->state.id_ = id;
+  self->state.accountant_ = accountant;
+  self->state.fs_ = fs;
+  self->state.path_ = store_path_for_partition(id);
+  self->request(self->state.fs_, caf::infinite, atom::mmap_v, self->state.path_)
+    .then(
+      [self](const chunk_ptr& chunk) {
+        if (auto table = read_parquet_buffer(chunk)) {
+          auto table_slices = std::vector<table_slice>{};
+          for (const auto& rb : arrow::TableBatchReader(*table)) {
+            if (!rb.ok())
+              self->send_exit(
+                self, caf::make_error(ec::format_error,
+                                      fmt::format("unable to read record "
+                                                  "batch: "
+                                                  "{} ",
+                                                  rb.status().ToString())));
+            table_slices.push_back(create_table_slice(*rb));
+            self->state.num_rows_ += (*rb)->num_rows();
+          }
+          self->state.table_slices_ = std::move(table_slices);
+        } else {
+          self->send_exit(self, table.error());
+        }
+        for (auto&& [query, rp] :
+             std::exchange(self->state.deferred_requests_, {})) {
+          VAST_TRACE("{} delegates {} (pending: {})", *self, query,
+                     rp.pending());
+          rp.delegate(static_cast<system::store_actor>(self), std::move(query));
+        }
+      },
+      [self](caf::error& err) {
+        VAST_ERROR("failed to read archive {}: {}", self->state.path_,
+                   to_string(err));
+        self->state.self_ = nullptr;
+      });
+  return {
+    [self](const query& query) -> caf::result<uint64_t> {
+      if (!self->state.table_slices_) {
+        auto rp = self->make_response_promise<uint64_t>();
+        self->state.deferred_requests_.emplace_back(query, rp);
+        return rp;
+      }
+      auto start = std::chrono::steady_clock::now();
+      auto slices = std::vector<table_slice>{};
+      auto num_hits = handle_lookup(self, query, *self->state.table_slices_);
+      duration runtime = std::chrono::steady_clock::now() - start;
+      auto id_str = fmt::to_string(query.id);
+      self->send(self->state.accountant_, "parquet-store.lookup.runtime",
+                 runtime,
+                 vast::system::metrics_metadata{{"query", id_str},
+                                                {"store-type", "passive"}});
+      self->send(self->state.accountant_, "parquet-store.lookup.hits",
+                 *num_hits,
+                 vast::system::metrics_metadata{{"query", id_str},
+                                                {"store-type", "passive"}});
+      return num_hits;
+    },
+    [self](atom::erase, const ids& xs) -> caf::result<uint64_t> {
+      auto num_rows = rank(xs);
+      VAST_ASSERT(num_rows == 0
+                  || num_rows
+                       == static_cast<unsigned long>(self->state.num_rows_));
```

Comment:
We have a `detail::narrow_cast` that wraps a `static_cast` and in debug mode checks if narrowing occurred.